contracts/root/StakeManager.sol

```diff
@@ -6,134 +6,194 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
+    address validator;
+    // add previous validators to priority queue
+    for (i = 0; i < validatorSet.validators.length; i++) {
+      validator = validatorSet.validators[i];
+      if (stakers[validator].status != ValidatorStatus.UNSTAKING) {
+        uint256 priority = _priority(validator, stakers[validator].amount, stakers[validator].data);
+        stakeQueue.insert(priority, stakingIdCount);
+        stakingIdToAddress[stakingIdCount] = validator;
+        stakingIdCount.add(1);
+      }
+    }
 
-  function unstake(uint256 amount, bytes data) public onlyStaker {
-    uint256 currentStake = totalStakedFor(msg.sender);
-    require(amount <= currentStake);
-
-    // reduce stake
-    uint256 stakedAmount = currentStake.sub(amount);
-    stakeList[stakers[msg.sender]].amount = stakedAmount;
+    require(stakeQueue.currentSize() >= _validatorThreshold); 
+    validatorSet = new ValidatorSet();
+    uint256 stakerId;
+      
+    for (uint256 i = 0; i < _validatorThreshold; i++) {
+      ( , stakerId) = stakeQueue.delMin();
+      validator = stakingIdToAddress[stakerId];
+      stakers[validator].status = ValidatorStatus.VALIDATOR;
+      validatorSet.addValidator(validator, stakers[validator].amount);
+      delete stakingIdToAddress[stakerId];
+    }
+    totalEpoch = currentEpoch.add(ValidatorSet.totalPower.div(
+      ValidatorSet.lowestPower).mul(_validatorThreshold));
+    
+    emit NewValidatorSet(_validatorThreshold, validatorSet.totalVotingPower, "0");
+  }
 
-    // reduce total stake
-    totalStake = totalStake.sub(amount);
+  // unstake and transfer amount for all valid exiters
+  function _unstake() private {
+    for (uint256 i = 0; i < exiterList.length; i++) {
+      address exiter = exiterList[i];
+      
+      if (stakers[exiter].status == ValidatorStatus.UNSTAKING && (
+        currentEpoch.sub(stakers[exiter].epoch)) <= minLockInPeriod ) {
+          
+        require(tokenObj.transfer(exiter, stakers[exiter].amount));
+        totalStake = totalStake.sub(stakers[exiter].amount);
+        emit Unstaked(exiter, stakers[exiter].amount, totalStake, "0");
+        delete stakers[exiter];
+
+        //delete from exiter list
+        exiterList[i] = exiterList[exiterList.length - 1]; 
+        delete exiterList[exiterList.length - 1]; 
+        // Todo: delete from staker list
+        // stakersList[] = stakersList[] delete index 
+      }
+    }
+  }
 
-    // transfer stake amount
-    tokenObj.transfer(msg.sender, amount);
+  function unstake(uint256 amount, bytes data) public onlyStaker { // onlyownder
+    // require(stakers[msg.sender]); //staker exists
+    require(stakers[msg.sender].epoch != 0); 
+    require(stakers[msg.sender].amount == amount);
+    
+    stakers[msg.sender].status = ValidatorStatus.UNSTAKING;
+    exiterList.push(msg.sender); 
+    
+    emit UnstakeInit(msg.sender, amount, totalStake.sub(amount), "0");
+  }
+  
+  function selectProposer() public onlyRootChain returns(address) {
```

Comment:
Can we change it to `getProposer()` and add `view` as it's public and just getter (not changing contract state)?

Reply:
It's changing the state via validatorset contract.